Sandbox builder and reduction counting script runner

lastcanal opened the pull request:
This implements a reduction counting script runner with an interface similar to `luerl:do/2`. I have been using this pattern to run untrusted lua scripts via luerl for my application. This pattern is best suited for short running scripts because it relies on limiting the number of reductions executed by the runner process.  

This method is able to stop 'fork bomb' scripts, eg:

```lua
x = 'a'; while true do x = x .. x end
```

It has the extra overhead of requiring a new process to be spawned to execute each request. The benefit  is we can easily control it's random state without dropping support for R17, as is the case with #84 . 

Here is an example of the full interface:

```erlang
Script = "x = 'a'; while true do x = x .. x end",
State = luerl:init(),
MaxReductions = 2000,
ProcessFlags = [{priority, low}],
RandomSeed = os:timestamp(),
{error, {reductions, R1}} =
    luerl_sandbox:run(Script, State, MaxReductions, ProcessFlags, RandomSeed)
```

I have also included a shortcut function for removing unsafe global tables and functions from an environment:

```erlang
luerl_sandbox:init(), %% default unsafe tables removed from new environment,
luerl_sandbox:init(luerl:init()), %% default unsafe tables removed from environment,
luerl_sandbox:init(luerl:init(), [ ['_G', math], ['_G', os, getenv] ]), %% custom tables removed from environment,

```

Here is my list of unsafe tables/functions that get removed by default:

```erlang
-define(SANDBOXED_GLOBALS, [
        [?LUERL_GLOBAL, io],
        [?LUERL_GLOBAL, file],
        [?LUERL_GLOBAL, os, getenv],
        [?LUERL_GLOBAL, package],
        [?LUERL_GLOBAL, load],
        [?LUERL_GLOBAL, loadfile],
        [?LUERL_GLOBAL, require],
        [?LUERL_GLOBAL, dofile],
        [?LUERL_GLOBAL, load],
        [?LUERL_GLOBAL, loadfile],
        [?LUERL_GLOBAL, loadstring]
    ]).
```

This is maybe best suited to be it's own separate application; let me know what you think!

run `cd examples/hello && make hello_sandbox` to try it out. 

```
erlc -I ../../src hello_sandbox.erl
erl -pa ../../ebin -s hello_sandbox run -s init stop -noshell
os.getenv with sandbox: {undef_function,<<"sandboxed">>}
catch error with os.getenv("HOME") with sandbox
killed process with reductions 2000 > 100
killed process with reductions 1999 > 100
Finished running with unlimited reductions 
random seed 7.0 == 7.0 != 75.0 
```

IgorKarymov commented:
Dear @rvirding any chance to merge this PR? It looks pretty usable for me too.

rvirding commented:
This looks interesting. Why are setting the random seed? If you could rebase it I will pull it.

lastcanal commented:
I've rebased on top of develop. I've also removed the random seed setting because it's redundant now with your latest changes. On top of that i've made the receive timeout configurable.
